src/hmacKey.ts

```diff
@@ -0,0 +1,355 @@
+
+export interface HmacKeyMetadata {
+  accessId: string;
+  etag?: string;
+  id?: string;
+  projectId?: string;
+  serviceAccountEmail?: string;
+  state?: string;
+  timeCreated?: string;
+  updated?: string;
+}
+
+export interface GetHmacKeyOptions extends GetConfig {
+  userProject?: string;
+}
+
+export interface UpdateHmacKeyOptions {
+  userProject?: string;
+}
+
+export interface UpdateHmacKeyMetadata {
+  state?: 'ACTIVE' | 'INACTIVE';
+  etag?: string;
+}
+
+export interface ToggleHmacKeyOptions {
+  etag?: string;
+}
+
+export interface HmacKeyMetadataCallback {
+  (err: Error | null, metadata?: HmacKeyMetadata, apiResponse?: Metadata): void;
+}
+
+export type HmacKeyMetadataResponse = [HmacKeyMetadata, Metadata];
+
+/**
+ * An HmacKey object contains metadata of an HMAC key created from a
+ * service account through the {@link Storage} client using
+ * {@link Storage#createHmacKey}.
+ *
+ * @class
+ */
+export class HmacKey extends ServiceObject<HmacKeyMetadata> {
+  accessId: string;
+  metadata: HmacKeyMetadata;
+  parent: Storage;
+
+  /**
+   * Constructs an HmacKey object.
+   *
+   * Note: this only create a local reference to an HMAC key, to create
+   * an HMAC key, use {@link Storage#createHmacKey}.
+   *
+   * @param {Storage} storage The Storage instance this HMAC key is
+   *     attached to.
+   * @param {string} accessId The unique accessId for this HMAC key.
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const hmacKey = storage.hmacKey('access-id');
+   */
+  constructor(storage: Storage, accessId: string) {
+    if (!accessId) {
+      throw new Error('An access ID is needed to create an HmacKey object.');
+    }
+
+    const methods = {
+      /**
+       * @typedef {object} DeleteHmacKeyOptions
+       * @property {string} [userProject] This parameter is currently ignored.
+       */
+      /**
+       * @typedef {array} DeleteHmacKeyResponse
+       * @property {object} 0 The full API response.
+       */
+      /**
+       * @callback DeleteHmacKeyCallback
+       * @param {?Error} err Request error, if any.
+       * @param {object} apiResponse The full API response.
+       */
+      /**
+       * Deletes an HMAC key.
+       * Key state must be set to `INACTIVE` prior to deletion.
+       * Caution: HMAC keys cannot be recovered once you delete them.
+       *
+       * The authenticated user must have `storage.hmacKeys.delete` permission for the project in which the key exists.
+       *
+       * @method HmacKey#delete
+       * @param {DeleteHmacKeyOptions} [options] Configuration options.
+       * @param {DeleteHmacKeyCallback} [callback] Callback function.
+       * @returns {Promise<DeleteHmacKeyResponse>}
+       *
+       * @example
+       * const {Storage} = require('@google-cloud/storage');
+       * const storage = new Storage();
+       *
+       * //-
+       * // Delete HMAC key after making the key inactive.
```

Comment:
Is this behavior that we could do by default, or at least optionally?

Reply:
Are you suggesting something like, check if the key is active, if it is then make it inactive then delete when calling `hmacKey.delete`?

Reply:
Yep!

Reply:
Decided against adding the force delete logic to prevent unintended deletion of key. Users need to explicitly make the key inactive first before deleting in case a key is still being used somewhere.